test/common.py

```diff
@@ -2567,52 +2609,65 @@ def browser_open(cls, url):
+      #   -a <exe_name>: The path to the executable to open, in this case Safari
+      browser_args = ['open', '--new', '--fresh', '--background', '-a'] + browser_args
+
     logger.info('Launching browser: %s', str(browser_args))
 
-    if WINDOWS and is_firefox():
-      cls.launch_browser_harness_windows_firefox(worker_id, config, browser_args, url)
+    if (WINDOWS and is_firefox()) or is_safari():
+      cls.launch_browser_harness_with_proc_snapshot_workaround(parallel_harness, config, browser_args, url)
     else:
       cls.browser_procs = [subprocess.Popen(browser_args + [url])]
 
   @classmethod
-  def launch_browser_harness_windows_firefox(cls, worker_id, config, browser_args, url):
-    ''' Dedicated function for launching browser harness on Firefox on Windows,
-    which requires extra care for window positioning and process tracking.'''
+  def launch_browser_harness_with_proc_snapshot_workaround(cls, parallel_harness, config, browser_args, url):
+    ''' Dedicated function for launching browser harness in scenarios where
+    we need to identify the launched browser processes via a before-after
+    subprocess snapshotting delta workaround.'''
 
+    # In order for this to work, each browser needs to be launched one at a time
+    # so that we know which process belongs to which browser.
     with FileLock(browser_spawn_lock_filename) as count:
-      # Firefox is a multiprocess browser. On Windows, killing the spawned
-      # process will not bring down the whole browser, but only one browser tab.
-      # So take a delta snapshot before->after spawning the browser to find
-      # which subprocesses we launched.
-      if worker_id is not None:
+      # Take a snapshot before spawning the browser to find which processes
+      # existed before launching the browser.
+      if parallel_harness or is_safari():
         procs_before = list_processes_by_name(config.executable_name)
+
+      # Browser launch
       cls.browser_procs = [subprocess.Popen(browser_args + [url])]
-      # Give Firefox time to spawn its subprocesses. Use an increasing timeout
-      # as a crude way to account for system load.
-      if worker_id is not None:
+
+      # Give the browser time to spawn its subprocesses. Use an increasing
+      # timeout as a crude way to account for system load.
+      if parallel_harness or is_safari():
         time.sleep(2 + count * 0.3)
         procs_after = list_processes_by_name(config.executable_name)
+
+        # Take a snapshot again to find which processes exist after launching
+        # the browser. Then the newly launched browser processes are determined
+        # by the delta before->after.
+        cls.browser_procs = list(set(procs_after).difference(set(procs_before)))
+        if len(cls.browser_procs) == 0:
+          logger.warning('Could not detect the launched browser subprocesses. The test harness may not be able to close browser windows if a test hangs, and at harness exit.')
```

Comment:
Can this be an error instead?

Reply:
I can make it an error, but it is only so when running with EMTEST_BROWSER_AUTO_CONFIG enabled. (I think btw disabling EMTEST_BROWSER_AUTO_CONFIG might currently be broken and atm not possible)

If user adds a custom browser string, and EMTEST_BROWSER_AUTO_CONFIG is not enabled, then the browser might run in a tab on an existing browser, in which case this detection will not work.